### ISSUE TITLE
Chrome query performance issue with MEDIAN and STDDEV aggregators

### ISSUE THREAD

**Issue by agerard** [agershun/alasql#534]
Hi, I'm wondering if you are aware of problems with running alasql (aggregation) queries in Chrome, or if maybe there's a way to improve the performance. The query is:

```js
alasql('SELECT count(1) AS ct, min(_date) AS minDate, max(_date) AS maxDate,  min(selectedChem) AS minparam, max(selectedChem) AS maxparam, AVG(selectedChem) AS avgparam, MEDIAN(selectedChem) AS medparam, STDEV( selectedChem) AS sdevparam FROM ? WHERE  selectedChem is not null AND selectedChem != -9999  ORDER BY _date', [ resultSet ]);
```

The resultSet is an alasql result object with max rows (currently) just under 6000. It's much faster in Firefox - Chrome just crawls. The query that supplies the initial resultSet is not nearly so slow in Chrome as this aggregation query. Any suggestions would be appreciated!

**Comment by agershun** [comment]
@agerard This is not Chrome issue, but rather AlaSQL realization of MEDIAN and STDEV aggregators. Now AlaSQL recalculate whole MEDIAN for each row. We need to rewrite them. I will see, what we can do.

**Comment by agerard** [comment]
Thanks so much! We really appreciate this level of responsiveness and will give prominent mention to you guys when this project - an interactive OL3 map of water sampling data - is released (hopefully with improved performance).

**Comment by agershun** [comment]
@agerard Thank you for request. I rewrote aggregators. Now MEDIAN and other statistical aggregators are fast (by my earlier mistake they took about O(N^2/2), but now only O(N) as it should be).

May I ask you for two things:
* Test these aggregators (are they work faster and their results are ok)
* What other aggregators will be useful

**Comment by mathiasrw** [comment]
Looks awesome !

I have updated https://github.com/agershun/alasql/wiki/User-Defined-Functions

**Comment by agershun** [comment]
:+1:

**Comment by mathiasrw** [comment]
I saw this in a test

```
+alasql.aggr.MYMEDIAN = function(v,s,acc){
 +  // Init
 +  if(typeof acc.arr == 'undefined') {
 +    acc.arr = [v];
 +    return v; 
 +  // Pass
 +  } else {
 +    acc.arr.push(v);
 +    var p = acc.arr.sort();
 +    return p[(p.length/2)|0];     
 +  };
 +};
```

I was of the impression that the 3rd parameter was only the last value, but I see now that its an object. Is it always an object, or will it be whatever you return from the first call?

**Comment by agershun** [comment]
This is old version of aggregator. Now we use three parameters:value, accumulator, and flag: 
1 - for first line
2 - for second and lines
3 - after all lines



Отправлено с iPhone

**Comment by mathiasrw** [comment]
But is accumulator the response from last call or is it always an object?

**Comment by agerard** [comment]
re: solved please test 
I will gladly test, just need to clarify if these changes will already be found in http://cdn.jsdelivr.net/alasql/0.2/alasql.min.js or should I download a specific build? Using the cdn link unfortunately I don't see a significant performance difference yet, but have also seen browsers - particularly Chrome - cache scripts despite settings....

Also, should I wait a bit to test? Recent comments suggest ongoing work.

**Comment by mathiasrw** [comment]
I think the speed improvement is already implemented. 

>  will already be found in http://cdn.jsdelivr.net/alasql/0.2/alasql.min.js or should I download a specific build? 

Please download and save the last development version from https://raw.githubusercontent.com/agershun/alasql/develop/dist/alasql.js to test the changes :)

**Comment by agershun** [comment]
Everything is implemented in ```develop``` branch, and will be included in official 0.2.4 at the end of the month. I tested these custom aggregators it with 1000000 records.

**Comment by agerard** [comment]
Thanks, got the dev version - it does seem faster (MUCH faster in chrome), but now I get

TypeError: s is undefined / s.arr.push(v); / alasql.js (line 10049, col 3)

in firebug and this is breaking subsequent flow.

**Comment by agershun** [comment]
@agerard Sorry for the bug. Could you answer on questions below to understand the problem:

1. Is this the problem only STDDEV() or VAR() aggregators, right?
2. Does this problem occur in  Firefox only or in Chrome also?
3. Could you show the sample of problematic SQL query?

Unfortunately, I can work on it only tomorrow.

**Comment by agerard** [comment]
1. Only stdev - if I take that out it's fine
2. both
3. same query as first submitted:
alasql('SELECT count(1) AS ct, min(_date) AS minDate, max(_date) AS maxDate,  min(selectedChem) AS minparam, max(selectedChem) AS maxparam, AVG(selectedChem) AS avgparam, MEDIAN(selectedChem) AS medparam, STDEV( selectedChem) AS sdevparam FROM ? WHERE  selectedChem is not null AND selectedChem != -9999  ORDER BY _date', [ resultSet ]);

resultSet data to which aggregators are being applied looks like:
cond_top	39.18
do_top	7.24
fc_top	22
op_top	0.22
salinity_top	25.54
tc_top	22
tkn_top	1.36

**Comment by mathiasrw** [comment]
The error is in

````
// Standard deviation
alasql.aggr.VAR = function(v,s,stage){
	if(stage == 1) {
		return {arr:[v],sum:v};
	} else if(stage == 2) {
		s.arr.push(v);
		s.sum += v;
		return s;
	} else {
		var N = s.arr.length;
		var avg = s.sum / N;
		var std = 0;
		for(var i=0;i<N;i++) {
			std += (s.arr[i]-avg)*(s.arr[i]-avg);
		}
		std = std/(N-1);
		return std;
	}
};
```

The code looks good to me. 

Could it be an error caused by mixing AVG, MEDIAN and STDEV in the same request?

**Comment by agershun** [comment]
Unfortunately, I cannot test it now.Could you replace ```STDEV()``` with ```SQRT(VAR(...))```?

**Comment by agerard** [comment]
no, it doesn'ts with SQRT(VAR( ... either.
s is defined the first time through the stage 2 loop, as with STDEV, and s remains undefined the second time through, but once the exception has occurred the watched value of s changes from undefined to "((((typeof p['default']['do_top']==='undefined') === (typeof !(undefined)==='undefined')))&&(((p['default']['do_top'])!=((-(9999))))))"
if that helps....

**Comment by agerard** [comment]
Also, 
alasql('SELECT  STDEV('eb_top ') AS sdevparam FROM ? WHERE eb_top is not null AND eb_top != -9999 ORDER BY _date', [ resultSet ]);
fails as a standalone query with the same resultSet (and same error)

**Comment by agershun** [comment]
Worst case: I can not reproduce the problem.

Here is my test code:
```js
    var resultSet = [
      {_date:new Date("01.01.2016"), selectedChem:1},
      {_date:new Date("01.01.2015"), selectedChem:2},
      {_date:new Date("01.10.2015"), selectedChem:3},
      {_date:new Date("10.10.2015"), selectedChem:4},
    ];
    var res = alasql('SELECT count(1) AS ct, min(_date) AS minDate, max(_date) AS maxDate, min(selectedChem) AS minparam, max(selectedChem) AS maxparam, AVG(selectedChem) AS avgparam, MEDIAN(selectedChem) AS medparam, STDEV(selectedChem) AS sdevparam FROM ? WHERE selectedChem is not null AND selectedChem != -9999 ORDER BY _date', [ resultSet ]);
```
and the result:
```js
[ { ct: 4,
    minDate: 1420059600000,
    maxDate: 1451595600000,
    minparam: 1,
    maxparam: 4,
    avgparam: 2.5,
    medparam: 3,
    sdevparam: 1.2909944487358056 } ]
```

Here is [jsFiddle](http://jsfiddle.net/agershun/8smL2j3q/2/) with latest version of AlaSQL. Could you say, what should I change in the example to highlight the bug?

**Comment by mathiasrw** [comment]
> `alasql('SELECT STDEV('eb_top ') AS sdevparam FROM ? WHERE eb_top is not null AND eb_top != -9999 ORDER BY _date', [ resultSet ]);`

Fails because the first string is `SELECT STDEV(` followed by an attempt to place the javascript variable `eb_top` but a `+` was expected. See http://jsfiddle.net/8smL2j3q/3/

If you replace  `'eb_top '` with `eb_top` you get no error: See http://jsfiddle.net/8smL2j3q/4/

**Comment by agershun** [comment]
:+1:

**Comment by agerard** [comment]
@mathiasrw - the single quotes in the last query were just an artifact of rendering the variable from my working code - when I retested your query locally in my firebug console it still didn't work. But the good news is that after downloading alasql.js from the jsfiddle (unfortunately overwrote the previous dev download so can't compare what might have gone wrong) it now works with my code! So whether I got a transitional version, something went wrong with the earlier download, or browser held on to old code somehow, it's all good now! Thanks!
